### ISSUE TITLE
Generic type resolves to `Unknown` with nested restricted type variables

### ISSUE THREAD

**Issue by kennipj** [microsoft/pyright#4474]
When a class is generic over a restricted type variable, and one of those restrictions is a class that is also generic, then pyright is unable to resolve the provided type argument for the restricted class. See below code for an example:

```python
from typing import Generic, TypeVar


_T = TypeVar("_T")


class A:
    ...


class B(Generic[_T]):
    ...


_A_or_B = TypeVar("_A_or_B", A, B)


class C(Generic[_A_or_B]):
    ...


def some_func(some_field: C[B[str]]) -> None:  # No error here
    reveal_type(some_field)  # Type of "some_field" is "C[B[Unknown]]"

```

mypy in this case reports:
```
some_example.py:22: error: Value of type variable "_A_or_B" of "C" cannot be "B"  [type-var]
some_example.py:23: note: Revealed type is "some_example.C[some_example.B[builtins.str]]"
```

pyright is able to correctly resolve the type by setting `_A_or_B = TypeVar("_A_or_B", bound=A | B)`. This of course result in a wider type variable, and may not always be desired.

It's not quite clear to me why this should result in an error like mypy reports, but if this should be an error, then I'd expect pyright to also report an error. If not, then I'd expect pyright to correctly resolve the type.

Tested on pyright `1.1.289`, and mypy `0.991`.

**Comment by erictraut** [comment]
This is working as designed. `B` is a generic class that accepts a single type argument, but when you define `_A_or_B`, you are omitting the type argument for `B`. That means `_A_or_B` has two constraints: `A` and `B[Unknown]`. If you enable the `reportMissingTypeArgument` diagnostic rule, pyright will alert you to the missing type argument. Since TypeVar constraints cannot be generic, the proper constraint in this case as `B[Any]` (or `B[str]`, depending on your intent).

I don't know why mypy emits an error in this case. The type `B[str]` is type compatible with `B[Any]`, so `B[str]` is a valid type argument for `C`. This looks to me like a bug in mypy.
